darwin: propagate build environment variables for Autotools

sethrj opened the pull request:
Both OpenMPI and GraphViz show failures due to overriding the default autoconf build environment variables. We may have to do something similar for other packages as they appear... alternatively, once libtool fixes their macOS build we could just autoreconf affected packages?

- https://debbugs.gnu.org/cgi/bugreport.cgi?bug=44605
- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97865
- https://trac.macports.org/ticket/61584

Perhaps we should just take the approach of patching libtool (like homebrew https://github.com/Homebrew/formula-patches/commit/e5fbd46a25e35663059296833568667c7b572d9a ) and forcing autoreconf on bigsur+ ?

hppritcha commented:
Looks good to me, although not able to test it myself.

trws commented:
What about adding it as a `@before` or `@after` for `setup_build_environment`?  That might let us add to it rather than override it completely, should catch most of these without having to find it in every place it comes up.  In fact, `@before` would have the pleasant side-effect of letting users override the setting in a package if that's necessary for some reason.

And now I think about it further and think perhaps we can't because those are for build phases rather than methods, that's really the effect we want though isn't it?

sethrj commented:
> And now I think about it further and think perhaps we can't because those are for build phases rather than methods, that's really the effect we want though isn't it?

Yep, unfortunately the idea won't work: `setup_build_environment` is a regular function called by `spack.build_environment` before running.

I think the `libtools` patching plan is more sensible, but forcing the deployment target is definitely more expedient.

sethrj commented:
@spackbot re-run pipeline

sethrj commented:
@trws Would you mind checking this out as well? Or do you think we should go with the option of patching `libtool` and forcing autoreconf on `bigsur`/`monterey`?

sethrj commented:
Hi @trws , I've built out a couple of Mac tool chains with this patch so I think it's at least a step in the right direction. Can you approve? Thanks!

trws reviewed (dismissed):
This works for me.  We should probably document that this is what's expected behavior, why we don't require this kind of thing so we can put functionality in the bases I'm not sure, but it seems now like a mistake.

sethrj committed:
darwin: propagate build environment variables for Autotools

GraphViz has failures due to overriding the default autoconf build
environment variables.

sethrj commented:
@trws Looks like the `openmpi` environment variable modification got deleted upstream -- I've rebased on master (but the gitlab CI is currently flaking out for unrelated reasons). Sorry to bother you for the reapproval.

sethrj commented:
@spackbot re-run pipeline

sethrj commented:
@trws Sorry but I had to rebase again to get off the broken develop :(

sethrj commented:
Fourth time's the charm 😂  thanks @trws !
